### ISSUE TITLE
Chap7-1.upynbのコード修正ポイント

### ISSUE THREAD

**Issue by umi-tyaahan** [koshian2/MosaicDeeplearningBook#10]
tensorflow 2.6.0で動かす場合、いくつか修正するとうまく動作しました。(100epochクリア）
参考に修正ポイントを記載します。

# 修正ポイント

## Q6 Hinge Loss
本文124pの参考回答例ですと、何か次元が合わない的なエラーだったので修正。
（詳細なエラー内容メモし忘れました…）

```python
    elif loss_type == "dis_real":
        # minval = tf.minimum(logits - 1, tf.zeros(logits.shape, dtype=logits.dtype)) # 本文での参考回答例
        minval = tf.minimum(logits -1,tf.zeros(tf.shape(logits),dtype=logits.dtype)) # 修正
        return -tf.reduce_mean(minval,axis=(1,2,3))
    else:
        # minval = tf.minimum(-logits - 1, tf.zeros(logits.shape, dtype=logits.dtype)) # 本文での参考回答例
        minval = tf.minimum(-logits-1,tf.zeros(tf.shape(logits),dtype=logits.dtype)) # 修正
        return -tf.reduce_mean(minval,axis=(1,2,3))
```

## データセットの作成部分
batchのdrop_remainder=Trueを追加

```python
    # dataset
    train_set = tf.data.Dataset.from_tensor_slices((train_orig, train_mosaic)).map(lambda *args: [preprocess(a) for a in args])
    train_set = train_set.shuffle(1024).batch(batch_size,drop_remainder=True).prefetch(buffer_size=tf.data.experimental.AUTOTUNE)
    test_set = tf.data.Dataset.from_tensor_slices((test_orig, test_mosaic)).map(lambda *args: [preprocess(a) for a in args])
    test_set = test_set.shuffle(1024).batch(batch_size,drop_remainder=True).prefetch(buffer_size=tf.data.experimental.AUTOTUNE)
```

# ご参考
tf.distribute.experimental.TPUStrategyはdeprecatedのようです。
代替としてtf.distribute.TPUStrategyを使うようです。
（切り替えたAPIでの動作はまだ試してないですが…🤔）

```python
# strategy = tf.distribute.experimental.TPUStrategy(tpu_cluster_resolver) # deprecated
strategy = tf.distribute.TPUStrategy(tpu_cluster_resolver) 
```

**Comment by koshian2** [comment]
@umi-tyaahan 
TF2.6の動作検証ありがとうございます。仕様変更が微妙にいやらしいですね。
このissueを演習問題にリンクさせていただきました。
